Microsoft.Toolkit.Uwp/Helpers/DispatcherQueueHelper.cs

```diff
@@ -0,0 +1,249 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading.Tasks;
+using Windows.System;
+
+namespace Microsoft.Toolkit.Uwp.Helpers
+{
+    /// <summary>
+    /// This class provides static methods helper for executing code in a DispatcherQueue.
+    /// </summary>
+    public static class DispatcherQueueHelper
+    {
+        /// <summary>
+        /// Extension method for <see cref="DispatcherQueue"/>. Offering an actual awaitable <see cref="Task"/> with optional result that will be executed on the given dispatcher.
+        /// </summary>
+        /// <param name="dispatcher">DispatcherQueue of a thread to run <paramref name="function"/>.</param>
+        /// <param name="function"> Function to be executed on the given dispatcher.</param>
+        /// <param name="priority">DispatcherQueue execution priority, default is normal.</param>
+        /// <returns>An awaitable <see cref="Task"/> for the operation.</returns>
+        /// <remarks>If the current thread has UI access, <paramref name="function"/> will be invoked directly.</remarks>
+        public static Task ExecuteOnUIThreadAsync(this DispatcherQueue dispatcher, Action function, DispatcherQueuePriority priority = DispatcherQueuePriority.Normal)
+        {
+            if (function is null)
+            {
+                throw new ArgumentNullException(nameof(function));
+            }
+
+            /* Run the function directly when we have thread access.
+             * Also reuse Task.CompletedTask in case of success,
+             * to skip an unnecessary heap allocation for every invocation. */
+
+            // Ignoring for now, but need to map the CurrentThreadID for all dispatcher queue code we have
+            /*
+            if (dispatcher.HasThreadAccess)
+            {
+                try
+                {
+                    function();
+
+                    return Task.CompletedTask;
+                }
+                catch (Exception e)
+                {
+                    return Task.FromException(e);
+                }
+            }
+            */
+
+            var taskCompletionSource = new TaskCompletionSource<object>();
+
+            bool ok = dispatcher.TryEnqueue(priority, () =>
```

Comment:
>ok [](start = 17, length = 2)

If ok is false, you probably want to return Task.CompletedTask - because the task never enqueued if this returns false, you'll never finish the await.  Or maybe you want to throw in that case?

This can return false during shutdown scenarios, for example, when the dispatcher is spinning down and can't process more work.  It's rare.

Reply:
Is this behavior different from CoreDispatcher? Our former helpers did this and it was never a problem. @michael-hawker , what do you think?

Reply:
If we're not going to use `ok` then probably just use the discard `_`

Reply:
Addressed by https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3206/commits/92f8328fd47b9731f239a161e037aa77956e6f3d